File: src/main/java/org/moinex/ui/dialog/investment/EditBondSaleController.java
/*
 * Filename: EditBondSaleController.java
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import javafx.fxml.FXML;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.model.investment.BondOperation;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.BondService;
import org.moinex.service.CategoryService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.service.WalletTransactionService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.moinex.util.enums.TransactionStatus;
import org.moinex.util.enums.TransactionType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public final class EditBondSaleController extends BaseBondTransactionManagement {
    @FXML private TextField netProfitField;
    @FXML private TextField grossYieldField;

    private BondOperation operation = null;

    @Autowired
    public EditBondSaleController(
            WalletService walletService,
            WalletTransactionService walletTransactionService,
            CategoryService categoryService,
            BondService bondService,
            I18nService i18nService) {
        super(walletService, walletTransactionService, categoryService, bondService, i18nService);
        this.i18nService = i18nService;
        transactionType = TransactionType.INCOME;
    }

    public void setOperation(BondOperation op) {
        this.operation = op;
        this.bond = op.getBond();

        String symbol = bond.getSymbol();
        bondNameLabel.setText(
                bond.getName() + (symbol != null && !symbol.isBlank() ? " (" + symbol + ")" : ""));

        setWalletComboBox(operation.getWalletTransaction().getWallet());

        descriptionField.setText(operation.getWalletTransaction().getDescription());
        unitPriceField.setText(operation.getUnitPrice().toString());
        quantityField.setText(operation.getQuantity().toString());

        if (operation.getFees() != null) {
            feesField.setText(operation.getFees().toString());
        }

        if (operation.getTaxes() != null) {
            taxesField.setText(operation.getTaxes().toString());
        }

        if (operation.getNetProfit() != null) {
            netProfitField.setText(operation.getNetProfit().toString());
        }

        statusComboBox.setValue(operation.getWalletTransaction().getStatus());
        categoryComboBox.setValue(operation.getWalletTransaction().getCategory());
        transactionDatePicker.setValue(operation.getWalletTransaction().getDate().toLocalDate());

        updateTotalPrice();
    }

    @FXML
    @Override
    protected void initialize() {
        super.initialize();

        unitPriceField
                .textProperty()
                .addListener(
                        (obs, oldVal, newVal) -> {
                            if (!netProfitField.isFocused()) {
                                calculateProfit();
                            }
                        });
        quantityField
                .textProperty()
                .addListener(
                        (obs, oldVal, newVal) -> {
                            if (!netProfitField.isFocused()) {
                                calculateProfit();
                            }
                        });
        feesField
                .textProperty()
                .addListener(
                        (obs, oldVal, newVal) -> {
                            if (!netProfitField.isFocused()) {
                                calculateProfit();
                            }
                        });
        taxesField
                .textProperty()
                .addListener(
                        (obs, oldVal, newVal) -> {
                            if (!netProfitField.isFocused()) {
                                calculateProfit();
                            }
                        });

        grossYieldField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.MONETARY_VALUE_REGEX)) {
                                grossYieldField.setText(oldValue);
                            } else {
                                if (!netProfitField.isFocused()) calculateProfit();
                                updateTotalPrice();
                                walletAfterBalance();
                            }
                        });

        netProfitField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.SIGNED_MONETARY_VALUE_REGEX)) {
                                netProfitField.setText(oldValue);
                            } else {
                                updateNetProfitColor(newValue);
                                updateTotalPrice();
                                walletAfterBalance();
                            }
                        });
    }

    @FXML
    @Override
    protected void handleSave() {
        Wallet wallet = walletComboBox.getValue();
        String description = descriptionField.getText();
        TransactionStatus status = statusComboBox.getValue();
        Category category = categoryComboBox.getValue();
        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        String feesStr = feesField.getText();
        String taxesStr = taxesField.getText();
        LocalDate saleDate = transactionDatePicker.getValue();

        if (wallet == null
                || description == null
                || description.isBlank()
                || status == null
                || category == null
                || unitPriceStr == null
                || unitPriceStr.isBlank()
                || quantityStr == null
                || quantityStr.isBlank()
                || saleDate == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_MESSAGE));

            return;
        }

        try {
            BigDecimal unitPrice = new BigDecimal(unitPriceStr);
            BigDecimal quantity = new BigDecimal(quantityStr);
            BigDecimal fees =
                    (feesStr != null && !feesStr.isBlank()) ? new BigDecimal(feesStr) : null;
            BigDecimal taxes =
                    (taxesStr != null && !taxesStr.isBlank()) ? new BigDecimal(taxesStr) : null;

            String netProfitStr = netProfitField.getText();
            BigDecimal netProfit =
                    (netProfitStr != null && !netProfitStr.isBlank())
                            ? new BigDecimal(netProfitStr)
                            : null;

            boolean feesEqual =
                    (operation.getFees() == null && fees == null)
                            || (operation.getFees() != null
                                    && fees != null
                                    && operation.getFees().compareTo(fees) == 0);
            boolean taxesEqual =
                    (operation.getTaxes() == null && taxes == null)
                            || (operation.getTaxes() != null
                                    && taxes != null
                                    && operation.getTaxes().compareTo(taxes) == 0);
            boolean netProfitEqual =
                    (operation.getNetProfit() == null && netProfit == null)
                            || (operation.getNetProfit() != null
                                    && netProfit != null
                                    && operation.getNetProfit().compareTo(netProfit) == 0);

            if (operation.getWalletTransaction().getWallet().getId().equals(wallet.getId())
                    && operation.getWalletTransaction().getDescription().equals(description)
                    && operation.getWalletTransaction().getStatus().equals(status)
                    && operation
                            .getWalletTransaction()
                            .getCategory()
                            .getId()
                            .equals(category.getId())
                    && operation.getUnitPrice().compareTo(unitPrice) == 0
                    && operation.getQuantity().compareTo(quantity) == 0
                    && feesEqual
                    && taxesEqual
                    && netProfitEqual
                    && operation.getWalletTransaction().getDate().toLocalDate().equals(saleDate)) {
                WindowUtils.showInformationDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_NO_CHANGES_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_NO_CHANGES_SALE_MESSAGE));
            } else {
                bondService.updateOperation(
                        operation.getId(),
                        wallet.getId(),
                        quantity,
                        unitPrice,
                        saleDate,
                        fees,
                        taxes,
                        netProfit,
                        category,
                        description,
                        status);

                WindowUtils.showSuccessDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_SALE_UPDATED_TITLE),
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_SALE_UPDATED_MESSAGE));
            }

            Stage stage = (Stage) bondNameLabel.getScene().getWindow();
            stage.close();
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_INVALID_NUMBER_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_INVALID_NUMBER_MESSAGE));
        } catch (EntityNotFoundException | IllegalArgumentException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_ERROR_UPDATING_SALE_TITLE),
                    e.getMessage());
        }
    }

    @Override
    protected void updateTotalPrice() {
        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        String netProfitStr = netProfitField.getText();

        BigDecimal totalPrice = new BigDecimal("0.00");

        if (unitPriceStr == null
                || quantityStr == null
                || unitPriceStr.isBlank()
                || quantityStr.isBlank()) {
            totalPriceLabel.setText(UIUtils.formatCurrency(totalPrice));
            return;
        }

        try {
            BigDecimal unitPrice = new BigDecimal(unitPriceStr);
            BigDecimal quantity = new BigDecimal(quantityStr);
            BigDecimal netProfit =
                    (netProfitStr != null && !netProfitStr.isBlank())
                            ? new BigDecimal(netProfitStr)
                            : BigDecimal.ZERO;

            totalPrice = unitPrice.multiply(quantity).add(netProfit);

            totalPriceLabel.setText(UIUtils.formatCurrency(totalPrice));
        } catch (NumberFormatException e) {
            totalPriceLabel.setText(UIUtils.formatCurrency(totalPrice));
        }
    }

    @Override
    protected void walletAfterBalance() {
        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        String netProfitStr = netProfitField.getText();
        Wallet wt = walletComboBox.getValue();

        if (unitPriceStr == null
                || unitPriceStr.isBlank()
                || quantityStr == null
                || quantityStr.isBlank()
                || wt == null) {
            UIUtils.resetLabel(walletAfterBalanceValueLabel);
            return;
        }

        try {
            BigDecimal unitPrice = new BigDecimal(unitPriceStr);
            BigDecimal quantity = new BigDecimal(quantityStr);
            BigDecimal netProfit =
                    (netProfitStr != null && !netProfitStr.isBlank())
                            ? new BigDecimal(netProfitStr)
                            : BigDecimal.ZERO;

            BigDecimal transactionValue = unitPrice.multiply(quantity).add(netProfit);

            if (transactionValue.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.resetLabel(walletAfterBalanceValueLabel);
                return;
            }

            BigDecimal walletAfterBalanceValue = wt.getBalance().add(transactionValue);

            if (walletAfterBalanceValue.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.setLabelStyle(
                        walletAfterBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                UIUtils.setLabelStyle(
                        walletAfterBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
            }

            walletAfterBalanceValueLabel.setText(UIUtils.formatCurrency(walletAfterBalanceValue));
        } catch (NumberFormatException e) {
            UIUtils.resetLabel(walletAfterBalanceValueLabel);
        }
    }

    private void updateNetProfitColor(String valueStr) {
        if (valueStr == null || valueStr.isBlank() || valueStr.equals("-")) {
            netProfitField
                    .getStyleClass()
                    .removeAll(
                            Constants.NEGATIVE_BALANCE_STYLE,
                            Constants.POSITIVE_BALANCE_STYLE,
                            Constants.NEUTRAL_BALANCE_STYLE);
            return;
        }

        try {
            BigDecimal value = new BigDecimal(valueStr);

            netProfitField
                    .getStyleClass()
                    .removeAll(
                            Constants.NEGATIVE_BALANCE_STYLE,
                            Constants.POSITIVE_BALANCE_STYLE,
                            Constants.NEUTRAL_BALANCE_STYLE);

            if (value.compareTo(BigDecimal.ZERO) > 0) {
                netProfitField.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);
            } else if (value.compareTo(BigDecimal.ZERO) < 0) {
                netProfitField.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                netProfitField.getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);
            }
        } catch (NumberFormatException e) {
            netProfitField
                    .getStyleClass()
                    .removeAll(
                            Constants.NEGATIVE_BALANCE_STYLE,
                            Constants.POSITIVE_BALANCE_STYLE,
                            Constants.NEUTRAL_BALANCE_STYLE);
        }
    }

    private void calculateProfit() {
        if (bond == null || netProfitField == null || grossYieldField == null) {
            return;
        }

        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        String interestStr = grossYieldField.getText();
        String feesStr = feesField.getText();
        String taxesStr = taxesField.getText();

        if (unitPriceStr == null
                || unitPriceStr.isBlank()
                || quantityStr == null
                || quantityStr.isBlank()) {
            netProfitField.setText("");
            return;
        }

        try {
            BigDecimal saleUnitPrice = new BigDecimal(unitPriceStr);
            BigDecimal quantity = new BigDecimal(quantityStr);
            BigDecimal interest = new BigDecimal(interestStr);
            BigDecimal fees =
                    (feesStr != null && !feesStr.isBlank())
                            ? new BigDecimal(feesStr)
                            : BigDecimal.ZERO;
            BigDecimal taxes =
                    (taxesStr != null && !taxesStr.isBlank())
                            ? new BigDecimal(taxesStr)
                            : BigDecimal.ZERO;

            BigDecimal averageUnitPrice = bondService.getAverageUnitPrice(bond);

            BigDecimal profit =
                    saleUnitPrice
                            .subtract(averageUnitPrice)
                            .multiply(quantity)
                            .add(interest)
                            .subtract(fees)
                            .subtract(taxes)
                            .setScale(2, RoundingMode.HALF_UP);

            netProfitField.setText(profit.toString());
        } catch (NumberFormatException e) {
            netProfitField.setText("");
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/update/EditBondSaleController.kt
/*
 * Filename: EditBondSaleController.kt (original filename: EditBondSaleController.java)
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isEqual
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.dto.WalletTransactionContextDTO
import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.investment.BondOperation
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.investment.base.BaseBondTransactionManagement
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.math.RoundingMode

@Controller
class EditBondSaleController(
    walletService: WalletService,
    categoryService: CategoryService,
    bondService: BondService,
    preferencesService: PreferencesService,
) : BaseBondTransactionManagement(
        walletService,
        categoryService,
        bondService,
        preferencesService,
    ) {
    @FXML
    private lateinit var netProfitField: TextField

    @FXML
    private lateinit var grossYieldField: TextField

    private lateinit var operation: BondOperation

    init {
        walletTransactionType = WalletTransactionType.INCOME
    }

    fun setOperation(op: BondOperation) {
        operation = op
        bond = op.bond

        val symbol = bond.symbol
        bondNameLabel.text = bond.name + if (!symbol.isNullOrBlank()) " ($symbol)" else ""

        setWalletComboBox(operation.walletTransaction!!.wallet)

        descriptionField.text = operation.walletTransaction!!.description
        unitPriceField.text = operation.unitPrice.toString()
        quantityField.text = operation.quantity.toString()
        feesField.text = operation.fees.toString()
        taxesField.text = operation.taxes.toString()
        netProfitField.text = operation.netProfit.toString()

        statusComboBox.value = operation.walletTransaction!!.status
        categoryComboBox.value = operation.walletTransaction!!.category
        transactionDatePicker.value = operation.walletTransaction!!.date.toLocalDate()
        includeInAnalysisCheckBox.isSelected = operation.walletTransaction!!.includeInAnalysis

        updateTotalPrice()
    }

    @FXML
    override fun initialize() {
        super.initialize()

        unitPriceField.textProperty().addListener { _, _, _ ->
            if (!netProfitField.isFocused) {
                calculateProfit()
            }
        }

        quantityField.textProperty().addListener { _, _, _ ->
            if (!netProfitField.isFocused) {
                calculateProfit()
            }
        }

        feesField.textProperty().addListener { _, _, _ ->
            if (!netProfitField.isFocused) {
                calculateProfit()
            }
        }

        taxesField.textProperty().addListener { _, _, _ ->
            if (!netProfitField.isFocused) {
                calculateProfit()
            }
        }

        grossYieldField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                grossYieldField.text = oldValue
            } else {
                if (!netProfitField.isFocused) calculateProfit()
                updateTotalPrice()
                walletAfterBalance()
            }
        }

        netProfitField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.SIGNED_MONETARY_VALUE_REGEX))) {
                netProfitField.text = oldValue
            } else {
                updateNetProfitColor(newValue)
                updateTotalPrice()
                walletAfterBalance()
            }
        }
    }

    @FXML
    override fun handleSave() {
        val wallet = walletComboBox.value
        val description = descriptionField.text
        val status = statusComboBox.value
        val category = categoryComboBox.value
        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val feesStr = feesField.text
        val taxesStr = taxesField.text
        val saleDate = transactionDatePicker.value

        if (wallet == null ||
            description.isNullOrBlank() ||
            status == null ||
            category == null ||
            unitPriceStr.isNullOrBlank() ||
            quantityStr.isNullOrBlank() ||
            saleDate == null
        ) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val unitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()
            val fees = if (!feesStr.isNullOrBlank()) feesStr.toBigDecimal() else BigDecimal.ZERO
            val taxes = if (!taxesStr.isNullOrBlank()) taxesStr.toBigDecimal() else BigDecimal.ZERO

            val netProfitStr = netProfitField.text
            val netProfit = if (!netProfitStr.isNullOrBlank()) netProfitStr.toBigDecimal() else BigDecimal.ZERO

            val feesEqual = operation.fees.isEqual(fees)
            val taxesEqual = operation.taxes.isEqual(taxes)
            val netProfitEqual = operation.netProfit.isEqual(netProfit)

            if (operation.walletTransaction!!.wallet.id == wallet.id &&
                operation.walletTransaction!!.description == description &&
                operation.walletTransaction!!.status == status &&
                operation.walletTransaction!!.category.id == category.id &&
                operation.walletTransaction!!.includeInAnalysis == includeInAnalysisCheckBox.isSelected &&
                operation.unitPrice.isEqual(unitPrice) &&
                operation.quantity.isEqual(quantity) &&
                feesEqual &&
                taxesEqual &&
                netProfitEqual &&
                operation.walletTransaction!!.date.toLocalDate() == saleDate
            ) {
                WindowUtils.showInformationDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_CHANGES_TITLE),
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_CHANGES_SALE_MESSAGE),
                )
            } else {
                operation.quantity = quantity
                operation.unitPrice = unitPrice
                operation.fees = fees
                operation.taxes = taxes
                operation.netProfit = netProfit

                bondService.updateBondOperation(
                    operation,
                    WalletTransactionContextDTO(
                        wallet,
                        saleDate.atStartOfDay(),
                        category,
                        description,
                        status,
                        includeInAnalysisCheckBox.isSelected,
                    ),
                )

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_SALE_UPDATED_TITLE),
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_SALE_UPDATED_MESSAGE),
                )
            }

            (bondNameLabel.scene.window as Stage).close()
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_INVALID_NUMBER_TITLE),
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_INVALID_NUMBER_MESSAGE),
                    )
                }
                is EntityNotFoundException, is IllegalArgumentException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_ERROR_UPDATING_SALE_TITLE),
                        e.message ?: "",
                    )
                }
                else -> throw e
            }
        }
    }

    override fun updateTotalPrice() {
        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val netProfitStr = netProfitField.text

        var totalPrice = BigDecimal("0.00")

        if (unitPriceStr.isNullOrBlank() || quantityStr.isNullOrBlank()) {
            totalPriceLabel.text = UIUtils.formatCurrency(totalPrice)
            return
        }

        runCatching {
            val unitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()
            val netProfit = if (!netProfitStr.isNullOrBlank()) netProfitStr.toBigDecimal() else BigDecimal.ZERO

            totalPrice = unitPrice.multiply(quantity).add(netProfit)

            totalPriceLabel.text = UIUtils.formatCurrency(totalPrice)
        }.onFailure {
            totalPriceLabel.text = UIUtils.formatCurrency(totalPrice)
        }
    }

    override fun walletAfterBalance() {
        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val netProfitStr = netProfitField.text
        val wt = walletComboBox.value

        if (unitPriceStr.isNullOrBlank() || quantityStr.isNullOrBlank() || wt == null) {
            UIUtils.resetLabel(walletAfterBalanceValueLabel)
            return
        }

        runCatching {
            val unitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()
            val netProfit = if (!netProfitStr.isNullOrBlank()) netProfitStr.toBigDecimal() else BigDecimal.ZERO

            val transactionValue = unitPrice.multiply(quantity).add(netProfit)

            if (transactionValue < BigDecimal.ZERO) {
                UIUtils.resetLabel(walletAfterBalanceValueLabel)
                return
            }

            val walletAfterBalanceValue = wt.balance.add(transactionValue)

            if (walletAfterBalanceValue < BigDecimal.ZERO) {
                UIUtils.setLabelStyle(walletAfterBalanceValueLabel, Styles.NEGATIVE_BALANCE_STYLE)
            } else {
                UIUtils.setLabelStyle(walletAfterBalanceValueLabel, Styles.NEUTRAL_BALANCE_STYLE)
            }

            walletAfterBalanceValueLabel.text = UIUtils.formatCurrency(walletAfterBalanceValue)
        }.onFailure {
            UIUtils.resetLabel(walletAfterBalanceValueLabel)
        }
    }

    private fun updateNetProfitColor(valueStr: String) {
        if (valueStr.isBlank() || valueStr == "-") {
            netProfitField.styleClass.removeAll(
                Styles.NEGATIVE_BALANCE_STYLE,
                Styles.POSITIVE_BALANCE_STYLE,
                Styles.NEUTRAL_BALANCE_STYLE,
            )
            return
        }

        runCatching {
            val value = valueStr.toBigDecimal()

            netProfitField.styleClass.removeAll(
                Styles.NEGATIVE_BALANCE_STYLE,
                Styles.POSITIVE_BALANCE_STYLE,
                Styles.NEUTRAL_BALANCE_STYLE,
            )

            when {
                value > BigDecimal.ZERO -> netProfitField.styleClass.add(Styles.POSITIVE_BALANCE_STYLE)
                value < BigDecimal.ZERO -> netProfitField.styleClass.add(Styles.NEGATIVE_BALANCE_STYLE)
                else -> netProfitField.styleClass.add(Styles.NEUTRAL_BALANCE_STYLE)
            }
        }.onFailure {
            netProfitField.styleClass.removeAll(
                Styles.NEGATIVE_BALANCE_STYLE,
                Styles.POSITIVE_BALANCE_STYLE,
                Styles.NEUTRAL_BALANCE_STYLE,
            )
        }
    }

    private fun calculateProfit() {
        if (!::netProfitField.isInitialized || !::grossYieldField.isInitialized) {
            return
        }

        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val interestStr = grossYieldField.text
        val feesStr = feesField.text
        val taxesStr = taxesField.text

        if (unitPriceStr.isNullOrBlank() || quantityStr.isNullOrBlank()) {
            netProfitField.text = ""
            return
        }

        runCatching {
            val saleUnitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()
            val interest = interestStr.toBigDecimal()
            val fees = if (!feesStr.isNullOrBlank()) feesStr.toBigDecimal() else BigDecimal.ZERO
            val taxes = if (!taxesStr.isNullOrBlank()) taxesStr.toBigDecimal() else BigDecimal.ZERO

            val averageUnitPrice = bondService.getAverageUnitPrice(bond)

            val profit =
                saleUnitPrice
                    .subtract(averageUnitPrice)
                    .multiply(quantity)
                    .add(interest)
                    .subtract(fees)
                    .subtract(taxes)
                    .setScale(2, RoundingMode.HALF_UP)

            netProfitField.text = profit.toString()
        }.onFailure {
            netProfitField.text = ""
        }
    }
}
